_pytest/config.py

```diff
@@ -1257,9 +1257,13 @@ def get_common_ancestor(args):
 
 
 def get_dirs_from_args(args):
-    return [d for d in (py.path.local(x) for x in args
-                        if not str(x).startswith("-"))
-            if d.exists()]
+    possible_dirs = [
+        py.path.local(str(arg).split('::')[0])
+        for arg in args
+        if not str(arg).startswith('-')
+    ]
+
+    return [d for d in possible_dirs if d.exists()]
```

Comment:
I've updated the PR to use a generator expression, small functions and made it always return dirs for consistency. I was also able to get rid of a couple of redundant checks in `get_common_ancestor`.